### ISSUE TITLE
RFC: Consider component augmentation

### ISSUE THREAD

**Issue by levithomason** [Semantic-Org/Semantic-UI-React#403]
# Dropdown

> _Copied from #401 regarding the `icon` prop.  This conversation now applies to all Dropdown types_.
## Preliminary Findings

The Dropdown supports many variations, including rendering as a Button or a Menu Item.  Here is what I've gathered that helped clarify some things for me:
- There is only ever one icon
- Dropdowns can appear as other components like a _Button_ or a _Menu Item_
- The icon is either a Dropdown icon, Button icon, or Menu Item icon but only one at a time

---

![image](https://cloud.githubusercontent.com/assets/5067638/17745560/1bf44318-6461-11e6-8c03-5e768982108a.png)

![image](https://cloud.githubusercontent.com/assets/5067638/17745386/757ce800-6460-11e6-8f1a-e5d0fa06df2c.png)

![image](https://cloud.githubusercontent.com/assets/5067638/17745403/80ecd402-6460-11e6-832b-45211128a25b.png)

![image](https://cloud.githubusercontent.com/assets/5067638/17745416/89256bfc-6460-11e6-837c-a7621f6a330f.png)

![image](https://cloud.githubusercontent.com/assets/5067638/17745436/96ddf3d6-6460-11e6-8923-0f32e288e5a6.png)

![image](https://cloud.githubusercontent.com/assets/5067638/17745538/014bbbe0-6461-11e6-9c7b-26b0d94144ad.png)

![image](https://cloud.githubusercontent.com/assets/5067638/17745446/9ef17886-6460-11e6-86da-56c7b665f014.png)

![image](https://cloud.githubusercontent.com/assets/5067638/17745527/ee4a9bec-6460-11e6-9737-fea130ab9ae1.png)

![image](https://cloud.githubusercontent.com/assets/5067638/17745534/f44a8d5e-6460-11e6-8f01-01b19d957a8c.png)
## Brainstorming Solutions

Given the above, we may be able to use a single `icon` prop after all.  It would just be applied to as either a Dropdown, Button, or Menu Item icon.

Our offline work going on with Forms may introduce a pattern that is helpful to Dropdowns as well.  We may allow either component augmentation or specialized subcomponents in `Form.Field`s.  This will facilitate our shorthand props API for Forms.  Here are some of the ideas we are playing with:

``` jsx
<Form.Field control='input' type='password' label='Enter password:' />

<Form.Field control={Checkbox} label='I agree to the terms and conditions' />
// or
<Form.Checkbox label='I agree to the terms and conditions' />
```

``` html
<div class="field">
  <label>Enter password:</label>
  <input type='password' />
</div>

<div class="field">
  <Checkbox label='I agree to the terms and conditions' />
</div>
```

There is also a need library wide to explicitly set the element type for a component.  Example, you may want to render a `<Segment />` as an HTML `<section />`.  We are considering a library wide standardized prop such as:

``` jsx
<Segment />
<Segment elementType='section' />
```

``` html
<div class="ui segment"></div>
<section class="ui segment"></section>
```
## Applying it to the Dropdown

All of these ideas are ways of augmenting components.  I have not thought this idea through at all, but it seems like we might support something along the lines of:

``` jsx
<Dropdown icon='filter' options={...} />
<Dropdown.Button labeled icon='settings' options={...} />
<Dropdown.Item icon='dropdown' options={...} />

// or

<Dropdown icon='filter' options={...} />
<Button elementType={Dropdown} labeled icon='settings' options={...} />
<Menu.Item elementType={Dropdown} icon='dropdown' options={...} />
```

The top group `Dropdown.Item` is already taken, so we'd have to solve the name conflict.  

The latter group I think makes more sense.  Consider the Button, all the props are regular button props (labeled, icon, etc) then the extra props (options) are passed to the `Dropdown`.  This would result in a SUI dropdown being rendered with all the props/classes of both the Button and Dropdown.

As I write this, I'm thinking it might help users better understand the augmentation if we use the prop name `as`:

``` jsx
<Button as={Dropdown} labeled icon='settings' options={...} />
<Menu.Item as={Dropdown} icon='dropdown' options={...} />

// or

<Dropdown as={Button} labeled icon='settings' options={...} />
<Dropdown as={Menu.Item} icon='dropdown' options={...} />
```
# MenuItem

> Copied from https://github.com/TechnologyAdvice/stardust/pull/382#discussion_r75179898

Many other components and use cases have us considering the use of a library wide `as` prop:

``` jsx
<Segment as='section' />
//=> <section class="ui segment"></section>

<MenuItem as='a' />
//=> <a class="item></a>

<Dropdown as={Button} options={[...]} />
//=> <div class="ui button dropdown" />
```

Especially handy for `react-router` use with the MenuItem as it would allow passing props for both components:

``` jsx
import { Link } from 'react-router'

<MenuItem link as={Link} to='/home' activeClass='active' />
//=> <a class="item active"></a>
```

The `MenuItem` would consume its own props (`link`) and spread the `rest`.  The `as` prop here means the MenuItem would render like:

``` jsx
render() {
  const ElementType = getElementType(props) // new util, like getUnhandledProps()

  return <ElementType {...rest} /* other menu item props */ />
}
```

The `getElementType()` method would hold the logic for returning the value of the `as` prop, else if `props.link` return `a`, else `div`.  We could add more logic to how components are allowed to determine their element type as well.  This is good because all components would be consistent.

**Comment by layershifter** [comment]
Linking #382

**Comment by levithomason** [comment]
Thanks, I updated the title include all components.  I also copied over the relevant MenuItem comments to the description.

**Comment by jeffcarbs** [comment]
We started implementing a lightweight react wrapper around SUI before coming across stardust. This is one of the missing things that's keeping us from switching over to stardust entirely.

We've solved this issue exactly like your `MenuItem` example (we named our argument `el`, but `as` reads well also):
- For the component type, accept either:
  - the name as a string `<MenuItem el='a' />`
  - a component type `<MenuItem el={Link} />`
  - default depending on the component. Usually just `div` but e.g. for `<Button>` the default is `button`
- Pass any unhandled props to the `el`

In your form example, this would mean:

``` jsx
// Good 
<Form.Field as={Checkbox} label='I agree to the terms and conditions' />
// Bad
<Form.Checkbox label='I agree to the terms and conditions' />
```

A reason to advocate for this is it allows the library consumer to pass arbitrary components, not just the ones included in stardust. For example, if I implemented my own custom checkbox component I'd be able to use it in the first case. The `react-router` `Link` is another example of this.

**Comment by levithomason** [comment]
@jcarbo appreciate the comments.  I fully agree with your points.  We also already spread unhandled props on the root node, so this will come for free once the root node is dynamic.  I also like `el` as a prop name and will consider using it.

Did you guys have any prop conflicts?  I'm thinking of cases where two components take the same prop name, the first will consume it while the second may have needed it as well.  Or, duplicate classNames: `<Dropdown as={Button} />` will have classes `ui button ui dropdown`.  Did you run into any of this and if so, how were those handled?

I'd like to hear from @davezuko on this as well.  I want to start on implementation for this today if at all possible.

**Comment by levithomason** [comment]
@jcarbo also:

> This is one of the missing things that's keeping us from switching over to stardust entirely.

If this is one, what else?! :) This project is the culmination of the greatest SUI React ideas out there.  We don't want to create another SUI React project, this project is joining the disparate efforts into _the_ SUI React project.  I believe we'll have better features and better support if we're in it together.  If you've got any other features or patterns that are better than what you see here, please do share.

**Comment by jeffcarbs** [comment]
> Did you guys have any prop conflicts? I'm thinking of cases where two components take the same prop name, the first will consume it while the second may have needed it as well. 

We haven't run into this yet, but my initial thought would be having a prop (e.g. `elProps`) to explicitly pass props down. For example:

``` js
render() {
  const ElementType = getElementType(props) // new util, like getUnhandledProps()

  return <ElementType {...elProps, ...rest} /* other menu item props */ />
}

// Usage
<Dropdown el={Button} elProps={{ icon: 'search' }} icon='dropdown' />
```

The default behavior would still remain re: `rest` so you'd only need to use that extra prop to avoid collisions.

> Or, duplicate classNames: <Dropdown as={Button} /> will have classes ui button ui dropdown. Did you run into any of this and if so, how were those handled?

I don't think this would pose an issue, although if it did I could see extending the class name helper to unique the values.

> This is one of the missing things that's keeping us from switching over to stardust entirely.

Off the top of my head it's mainly just customization of some elements to fit our specific needs, although I'm sure we'd probably be able to just wrap these components if need be.  I'll try to think about this a bit more and get back to you with some concrete examples.

**Comment by levithomason** [comment]
Cool, we're on the same page again.  You can expect this to roll out soon.  Keep us posted on your findings.

**Comment by levithomason** [comment]
@jcarbo the `as` prop is now merged, released in `v0.37.0`.  All components can be augmented.  The final element type priority became:
1. User defined an `as` prop
2. `getElementType()` `typeMap` arg mapped a prop to an element type: `{ onClick: 'a' }`
3. Inferred type: `href => a` for now
4. `defaultProp.as`
5. `div`

See the lib function here: https://github.com/TechnologyAdvice/stardust/blob/master/src/lib/getElementType.js

See it on the docs here:
http://technologyadvice.github.io/stardust/introduction
